### PR TITLE
Add error code to execution

### PR DESCRIPTION
**Pull request by mitchell-conrad**

### CODE REVIEW

File: util/execution.hpp [hunk 1]
```diff
@@ -25,18 +26,33 @@ class Execution{
...
+
+            if(errors_.empty()) {
+                return 0;
+            } else {
+                return 1;
+            }
         };
 
+        auto GetErrors() -> std::vector<std::string> {
+            return errors_;
```
Comment (JacksonM8) [inline util/execution.hpp]: Should take lock here

File: util/execution.hpp [hunk 2]
```diff
@@ -25,18 +26,33 @@ class Execution{
             for(auto func : terminate_functions_){
                 func();
             }
+
+            if(errors_.empty()) {
```
Comment (JacksonM8) [inline util/execution.hpp]: Should take a lock here

File: util/execution.hpp [hunk 3]
```diff
@@ -14,9 +14,10 @@ class Execution{
         std::condition_variable lock_condition_;
 
         std::vector<std::function<void()> > terminate_functions_;
+        std::vector<std::string> errors_;
```
Comment (JacksonM8) [inline util/execution.hpp]: Could this possibly be a list of exceptions?

**Review by JacksonM8 — CHANGES_REQUESTED** [review]

**Review by danielrfraser — COMMENTED** [review]
I think given that this is a more Generic class, updating the interupt function signature to look somrthing like this:

void Interupt(int return_code = 0){
   if(return_code != 0){
    return_status = return_code;
 }
}
re_node_manager already prints the errors, and this would be a far cleaner way of dealing with the errors?

File: util/execution.hpp [hunk 4]
```diff
@@ -1,48 +1,71 @@
...
+#ifndef RE_COMMON_UTIL_EXECUTION
+#define RE_COMMON_UTIL_EXECUTION
 
-#include <mutex>
 #include <condition_variable>
 #include <functional>
-#include <vector>
 #include <iostream>
+#include <mutex>
+#include <numeric>
```
Comment (JacksonM8) [inline util/execution.hpp]:
```suggestion
```

File: util/execution.hpp [hunk 5]
```diff
@@ -1,48 +1,71 @@
...
 
-        bool IsInterrupted(){
-            std::unique_lock<std::mutex> lock(mutex_);
-            return terminated;
+            throw std::runtime_error(message);
         }
-        
+    }
+
+    void SetError(std::unique_ptr<std::exception> error)
```
Comment (JacksonM8) [inline util/execution.hpp]:
```suggestion
    void AddException(std::unique_ptr<std::exception> error)
```

**Review by JacksonM8 — CHANGES_REQUESTED** [review]

**Pull request merged**